File: src/main/java/org/idempiere/common/base/Service.java
package org.idempiere.common.base;

/**
 * This is a very simple factory for service locators
 * 
 * @author viola
 * 
 */
public abstract class Service {

	private static IServiceLocator theLocator;
	protected static Service instance;

	protected abstract IServiceLocator getLocator();
    private final static Object mutex = new Object();

	/**
	 * 
	 * @return service locator instance
	 */
	public static IServiceLocator locator() {
        synchronized (mutex) {
            if (theLocator != null) {
                return theLocator;
            } else if (instance != null) {
                theLocator = instance.getLocator();
                return theLocator;
            } else throw new IllegalStateException("Service Locator");
        }
	}
}

File: src/main/kotlin/org/idempiere/common/base/Service.kt
package org.idempiere.common.base

/**
 * This is a very simple factory for service locators
 *
 * @author viola
 */
abstract class Service {

    protected abstract val locator: IServiceLocator

    companion object {

        private var theLocator: IServiceLocator? = null
        internal var instance: Service? = null
        private val mutex = Any()

        /**
         *
         * @return service locator instance
         */
        fun locator(): IServiceLocator? {
            synchronized(mutex) {
                if (theLocator != null) {
                    return theLocator
                } else if (instance != null) {
                    theLocator = instance!!.locator
                    return theLocator
                } else
                    throw IllegalStateException("Service Locator")
            }
        }
    }
}